docs/architecture/adr-028-public-key-addresses.md

```diff
@@ -208,20 +208,29 @@ If we want to create an address for a module account depending on more than one
 btcAtomAMM := address.Module("amm", btc.Addrress() + atom.Address()})
 ```
 
-We can continue the derivation process and can create an address for a submodule account.
+#### Derived Addresses
 
-```
-func Submodule(address []byte, derivationKey []byte) {
-    return Hash("module", address + derivationKey)
+Module address is a special case of more general _derived_ address defined by the `Derive` function:
+
+```go
+func Derive(address []byte, derivationKey []byte) {
+    return Hash(addres, derivationKey)
 }
 ```
 
-NOTE: if `address` is not a hash based address (with `LEN` length) then we should use `LengthPrefix`. An alternative would be to use one `Module` function, which takes a slice of keys and mapped with `LengthPrefix`. For final version we need to validate what's the most common use.
-
 
 **Example**  For a cosmwasm smart-contract address we could use the following construction:
 ```
-smartContractAddr := Submodule(Module("cosmwasm", smartContractsNamespace), smartContractKey)
+smartContractAddr := Derived(Module("cosmwasm", smartContractsNamespace), []{smartContractKey})
+```
+
+We can also define a function which will derive an address based on multiple keys (path). The function is similar to the `NewComposed`, however it doesn't sort the derivation keys:
+
+```go
+func DeriveMulti(address []byte, derivationKeys [][]byte) {
+    keys = map(derivationKeys, \k -> LengthPrefix(k))
+    return Hash(LengthPrefix(address), keys[0] + ... + keys[n])
```

Comment:
It's a bit confusing for me that `Derive` doesn't length-prefix derivationKey, and `DeriveMulti` length-prefixes.

I understand `Derive` doesn't need legnth-prefixing. But isn't it just simpler to always use the same algorithm (i.e. always lenght-prefix), so that `Derive == DeriveMulti with 1 derivationKey`?

Reply:
Indeed, we use length prefixing only if we want to glue a sequence of keys. So, for `Derive` it's unnecessary operation.

Reply:
> So, for Derive it's unnecessary operation.

Yes, but does it create more confusion? We could just:

```go
func Derive(address []byte, derivationKey []byte) []byte {
    return DeriveMulti(address, []byte{derivationKey})
}
```

so that we have 1 algorithm only.

Reply:
I think the confusion will be there anyway: `DeriveMulti` doesn't decompose into multiple Derive calls (even with LengthPrfixing) - what's in the function comment:
```
// NOTE: DeriveMulti(addr, [k1, k2]) != Derive(Derive(addr, k1), k2)
//                                   != Derive(Derive(addr, k2), k1)
```
The motivation here was to use a composed derivation path, Normally we would put slashes into the segments. But here, the key can be anything, so we can't have any special separator.


So to summarize, we have 3 options:
1. leave as is
2. remove `DeriveMulti`, and not adding it later in the form described above.
3. use @amaurym  suggestion to use 

Maybe we don't need `DeriveMuli` - it seams that we will always have the `parent` key and will never derive into subusub-account without having a sub-account. What do you think?

Reply:
> // NOTE: DeriveMulti(addr, [k1, k2]) != Derive(Derive(addr, k1), k2)
> //                                   != Derive(Derive(addr, k2), k1)

Another option:

4. DeriveMulti is composition of Derive

```golang
func DeriveMulti(address []byte, derivationKeys [][]byte) {
  result := address
  for _, derivKey := range derivationKeys {
    result = Derive(result, derivKey)
  }
  return result
}
```

The 1st line of your `NOTE` is then an equality (less confusion), and we never use length-prefixing (also less confusion). The 2nd line stays an inequality, but in general compositions are never commutative.

> Maybe we don't need DeriveMuli - it seams that we will always have the parent key and will never derive into subusub-account without having a sub-account. What do you think?

True, that also makes sense. Overall I'm actually in favor of `Option 2: remove DeriveMulti` for now.

Reply:
Yes, that's why I was more thinking about the `Option 2`. Again, the motivation for `DeriveMulti` was not to use composition.

Reply:
I mean - the motivation was to have a sequence of parameters, which don't compose, but use the whole sequence as a single parameter (as we have in the hardware wallets now).